src/server/gc_worker/compaction_filter.rs

```diff
@@ -247,50 +303,39 @@ impl CompactionFilter for WriteCompactionFilter {
         _: &mut Vec<u8>,
         _: &mut bool,
     ) -> bool {
-        let (key_prefix, commit_ts) = match Key::split_on_ts_for(key) {
-            Ok((key, ts)) => (key, ts.into_inner()),
-            // Invalid MVCC keys, don't touch them.
-            Err(_) => return false,
-        };
+        self.near_seek_distance += 1;
+        let (key_prefix, commit_ts) = split_ts(key);
+        if commit_ts > self.safe_point {
+            return false;
+        }
 
+        self.versions += 1;
         if self.key_prefix != key_prefix {
             self.key_prefix.clear();
             self.key_prefix.extend_from_slice(key_prefix);
             self.remove_older = false;
+            self.handle_deleting();
             self.switch_key_metrics();
         }
 
-        if commit_ts > self.safe_point {
-            return false;
-        }
-
-        self.versions += 1;
         let mut filtered = self.remove_older;
-        let write = match WriteRef::parse(value) {
-            Ok(write) => write,
-            // Invalid MVCC keys, don't touch them.
-            Err(_) => return false,
-        };
-
+        let write = parse_write(value);
         if !self.remove_older {
-            // here `filtered` must be false.
             match write.write_type {
                 WriteType::Rollback | WriteType::Lock => filtered = true,
                 WriteType::Put => self.remove_older = true,
                 WriteType::Delete => {
                     self.remove_older = true;
-                    if self.bottommost_level {
-                        filtered = true;
-                        self.gc_before_delete_mark(key, key_prefix);
-                    }
+                    filtered = true;
+                    self.deleting.0 = key_prefix.to_vec();
```

Comment:
What's the rational to actively remove stale versions after Delete but not Put?

Reply:
For MVCC delete marks, it's necessary to delete all stale versions (maybe not involved in the compaction) older than them because the marks themselves are not kept.